### ISSUE TITLE
Allow publishing source on Etherscan

### ISSUE THREAD

**Issue by iamdefinitelyahuman** [eth-brownie/brownie#411]
### Overview
It should be possible (and simple) to verify a deployed contract's source code on etherscan.

### Specification
Not 100% on how this would be best implemented. Some thoughts:

1. Add an internal "flattener" to turn multiple source files into a single file. Might also be worth exposing this on it's own via the CLI.
2. Add a method to `Contract`, to be called from in the console or a script.
3. Consider a flag in `brownie run` to auto-verify all sources after deployment?

Possible syntax:

```python
contract = ContractContainer.deploy({'from': accounts[0]})
contract.publish_on_etherscan()
```
and / or
```python
ContractContainer.deploy({'from': accounts[0]}, publish=True)
```

**Comment by Skyge** [comment]
Does anyone work on this issue? if not, I would like to solve this issue.

**Comment by WyseNynja** [comment]
Etherscan updated to support verifying from solidity-standard-json-input.

The `publish_on_etherscan` function should send [a POST like this](https://etherscan.io/sourcecode-demo.html).

**Comment by Netherdrake** [comment]
What is the current best way to achieve this for contracts created trough `brownie`?

**Comment by fubuloubu** [comment]
I think it would be good practice to have a `publish_sources: True|False` key that pushes to etherscan (or whatever block explorer) whenever a contract is deployed to a public network.

**Comment by sabotagebeats** [comment]
The json seems located in `build/contracts/` but where can I find the hex encoded constructor abi within `build/deployments/` ? 

[https://github.com/eth-brownie/brownie/issues/766#issuecomment-711619114](url)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 0.5 ETH (236.29 USD @ $472.58/ETH)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://chat.gitcoin.co/'>Gitcoin Chat</a>
* $1,045,515.92 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by oogetyboogety** [comment]
Looks like some of the work may have already started for this, but is blocked, in https://github.com/eth-brownie/brownie/issues/766?

**Comment by banteg** [comment]
Maybe we should also consider a standard json format instead of flattening for Solidity?
https://kndrck.co/posts/verify-contracts-etherscan-100/

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 266 years from now.
Please review their action plans below:


**1) [oogetyboogety](https://gitcoin.co/oogetyboogety) has been approved to start work.**

1) Create the internal flattener with an accessible CLI optoin and some standard for arranging/flattening multiple source files
2) add the publish_on_etherscan method to the contract conforming to the specs in https://etherscan.io/sourcecode-demo.html
3) add a build step in the brownie smart contract building and testing pipeline
**2) [developerfred](https://gitcoin.co/developerfred) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?mutate_worker_action=approve&worker=developerfred) | [reject worker](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?mutate_worker_action=reject&worker=developerfred))_.**

My Plan and build this function in cli that self-identifies the contract models before delivering.  ex: contracts with experimental ABI and contracts with default ABI
**3) [hhio618](https://gitcoin.co/hhio618) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?mutate_worker_action=approve&worker=hhio618) | [reject worker](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?mutate_worker_action=reject&worker=hhio618))_.**

My plan:
1. Add a flattener to brownie using this PyPI package: solidity-flattener.
2. Add a method to Contract to verify a source code on etherscan.
3. Add --auto-verify flag to brownie run.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068).

**Comment by Macarse** [comment]
@oogetyboogety sorry for the delay, I didn't notice I had to approve someone to start the work.

**Comment by iamdefinitelyahuman** [comment]
#651 could also be useful here.  I think verifying by submitting the standard JSON compiler input will be less error-prone.

**Comment by oogetyboogety** [comment]
@banteg  I need to validate to make sure there are not more than 10 imports, right?

**Comment by oogetyboogety** [comment]
@banteg  how should the interface look for an etherscan API key? `solt` apparently uses some shared, potentially rate-limited API key[1] and I am trying to dig further into what this [2] means for brownie's use case of the verify & publish POST API:
`Someone using the tools from dapp.tools for instance, in conjunction with something like dapp-pm to manage their dependencies was out of luck until now, and I’ve tried to make it as easy as possible for users to get their deployed contracts verified with as little headache as possible. This means including all the API keys necessary so that (best case) you shouldn’t need to create, store, and manage your own Etherscan API key if you don’t want to, although you can do that as well.`
[1] https://github.com/hjubb/solt/blob/main/src/commonMain/kotlin/com/github/hjubb/solt/Verifier.kt#L79-L83
[2] https://blog.jubb.xyz/post/solt-release/

**Comment by Macarse** [comment]
I would read from env variable. Other tools use `ETHERSCAN_TOKEN` or at least that's what I have in my env vars.

**Comment by oogetyboogety** [comment]
gotcha, thanks

**Comment by iamdefinitelyahuman** [comment]
Brownie is already expecting the API key as an environment variable `ETHERSCAN_TOKEN` elsewhere in the code, so this is the best option.

https://github.com/eth-brownie/brownie/blob/b7c72534adfadd29c136b21cb175e30c6437f7c0/brownie/network/contract.py#L1370-L1379

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@oogetyboogety due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by oogetyboogety** [comment]
Should have this pr ready today

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@oogetyboogety due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by oogetyboogety** [comment]
Right now i'm recompiling every time. I don't see anywhere where the output json from solc is stored, so I was trying to generate it again like this:

```
>>        input_json = compiler.generate_input_json(self._project._sources._contract_sources)
>>        output_json = compiler.compile_from_input_json(input_json)
```

should we be storing this in case this has already run? I guess my question is should we be caching solc's output json somewhere, if this is not already taking place? I looked but I don't see it being cached anywhere currently, as far as I can tell.

**Comment by oogetyboogety** [comment]
Also, to avoid reversing the inputs of the constructor arguments used in the ContractContainer's deploy call, as described here, https://ethereum.stackexchange.com/a/65171, I've been saving the arguments used in the contract's constructor call into its state.

**Comment by oogetyboogety** [comment]
OK I think the etherscan API just takes the hex encoded input to the constructor in self.tx.input[2:]

Does this look right?
```
>>        post_data: Dict = {
                  "module": "contract",
                  "action": "verifysourcecode",
                  "codeformat": "solidity-standard-json-input",
                  "contractaddress": self.address,
                  "sourceCode": output_json,
                  "contractname": f"{output_json['contracts'].keys()[-1]}:{contract_name}",
                  "compilerversion": self._build['compiler']['version'],
                  "optimizationUsed": self._build['compiler']['optimizer']['enabled'],
                  "runs": self._build['compiler']['optimizer']['runs'],
                  "constructorArguments": self.tx.input[2:],
                  "evmversion": self._build['compiler']['evm_version'],
                  "licensetype": 1
              }
>>        for i, dependency in enumerate(self._build['dependencies']):
              if i <= 10:
                  post_data[f"libraryname{i}"] = dependency
                  if self._project[dependency][-1]:
>>                        post_data[f"libraryaddress{i}"] = self._project[library][-1].address[-40:]
              else:
                  raise Exception('Too many libraries to publish on explorer')
```

**Comment by oogetyboogety** [comment]
I seem to keep getting ValueError: Failed to verify sourcecode from API: Pending in queue in the PR

is there a paradigm you follow on backoff of requests?

Thanks,
oogety

**Comment by oogetyboogety** [comment]
https://d1.awsstatic.com/builderslibrary/pdfs/timeouts-retries-and-backoff-with-jitter.pdf?did=ba_card-body&trk=ba_card-body

**Comment by oogetyboogety** [comment]
I get an exception in the code, but I guess it validates?
```
>>> t = accounts[0].deploy(project.TokenProject.Token, "Test Token", "ETH", None, None)
Transaction sent: 0xbfd0a9910b6048040618a1081e2b978e089088fc63de65782cfb5184457f1dc3
  Gas price: 5.00002159 gwei   Gas limit: 504546   Nonce: 23
  Token.constructor confirmed - Block: 9192203   Gas used: 458679 (90.91%)
  Token deployed at: 0x67f502cafC15a9783CBce4899F3e6671DC706e1F

>>> t.publish_to_explorer()
Verifying source of 0x67f502cafC15a9783CBce4899F3e6671DC706e1F from api-ropsten.etherscan.io...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/code/brownie/brownie/network/contract.py", line 543, in publish_to_explorer
    raise ValueError(f"Failed to verify sourcecode from API: {data['result']}")
ValueError: Failed to verify sourcecode from API: Pending in queue
```
But if you check Etherscan, you'll see where I validated brownie's input json manually and it realizes it's verified:
 Contract Source Code Verified (Similar Match)
Note: This contract matches the deployed ByteCode of the Source Code for Contract 0xcb61b5ec070c2ecebd219

So, I've got a little bit of work to do to figure out what I'm doing wrong to get a Pending result from the API here. I may need to touch base with Etherscan folks in some way.

**Comment by Macarse** [comment]
@oogetyboogety rebase your branch, you are missing a lot of changes.

**Comment by oogetyboogety** [comment]
@Macarse think i fixed the conflicts now

**Comment by oogetyboogety** [comment]
Also need to fix the hardcoded license i'm using in the API call

**Comment by Macarse** [comment]
@oogetyboogety you are still behind. also plz run lint so the PR checks pass.

**Comment by oogetyboogety** [comment]
@Macarse oh gotcha let me do that now

**Comment by oogetyboogety** [comment]
@Macarse  fixed

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@oogetyboogety due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by oogetyboogety** [comment]
@gitcoinbot i'll take another look at getting the api calls right this weekend

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by eyooooo** [comment]
hey guys just wanted to write a workaround i use for this. when i deploy with brownie i write a csvfile to disk and then i have a bash script that feeds the rows into `solt` for auto verify. it works like 50% of the time, solt doesnt always succeed. hope this helps.

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@oogetyboogety due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by Macarse** [comment]
@oogetyboogety plz check https://github.com/eth-brownie/brownie/pull/896
Can you create the etherscan api integration?

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@oogetyboogety due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by oogetyboogety** [comment]
@Mecarse  I can definitely try the flattened way.

I'm trying to get it working with the API using solidity-json-output, as you can see in my reference PR. I think I might be close,  I just need to figure out how to properly use the API.

**Comment by oogetyboogety** [comment]
I'll do that today

**Comment by oogetyboogety** [comment]
Think I got this working based on what I learned from https://github.com/rkalis/truffle-plugin-verify/blob/master/verify.js, will be updating in the next couple days with linted PR again

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by freightledger** [comment]
So close 🥂

**Comment by gitcoinbot** [comment]
@oogetyboogety Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=1) | [3 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=3) | [5 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=5) | [10 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=10) | [100 days](https://gitcoin.co/issue/eth-brownie/brownie/411/100024068?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__The funding of 0.5 ETH  (365.14 USD @ $730.28/ETH) attached to this issue has been **cancelled** by the bounty submitter__

 
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://chat.gitcoin.co/'>Gitcoin Chat</a>
* $1,006,745.79 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by Macarse** [comment]
@oogetyboogety I canceled the bounty after your inactivity.

**Comment by oogetyboogety** [comment]
Was still working on getting this working, having some trouble getting this working for 100% of contracts. I wonder if anyone had a chance to look at the pr i had in progress

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by fubuloubu** [comment]
This one is in progress, really solid approach I think will probably be what ends up being merged:
https://github.com/eth-brownie/brownie/pull/914

**Comment by oogetyboogety** [comment]
Oh gotcha. If it behooves anyone to look I made a PR in progress where the approach might be less solid it uses the solidity json file as opposed to flattening a contract. I was actually trying to get this to cooperate on all test contracts in brownie with the Etherscan API on Ropsten, without exceeding the request limits Etherscan API imposes while testing. If you would like for me to resume work on this, let me know.

**Comment by Netherdrake** [comment]
I've had issues getting contracts verified with the flattening approach. Really hopeful to see if @oogetyboogety can figure it out.

**Comment by matnad** [comment]
Implemented in #914. See PR for details.

**Comment by oogetyboogety** [comment]
`Make sure all your source files use the same compiler version, otherwise the verification will fail.`
I know the bounty is cancelled, but in case any one cares to look into this further, I think this might be where I was running into issues.

If we have linked contracts at already deployed addresses, it would make sense to get the error I kept getting if they were compiled using two different solidity commit hashes or versions.

FYI for future reference.
